src/lib/operators/join_sort_merge/radix_cluster_sort.hpp

```diff
@@ -8,46 +8,42 @@
 #include <utility>
 #include <vector>
 
+#include <boost/sort/sort.hpp>
+
 #include "column_materializer.hpp"
 #include "hyrise.hpp"
 #include "resolve_type.hpp"
 #include "utils/timer.hpp"
 
 namespace opossum {
 
-/**
-* The RadixClusterOutput holds the data structures that belong to the output of the clustering stage.
-*/
+// The RadixClusterOutput holds the data structures that belong to the output of the clustering stage.
 template <typename T>
 struct RadixClusterOutput {
-  std::unique_ptr<MaterializedSegmentList<T>> clusters_left;
-  std::unique_ptr<MaterializedSegmentList<T>> clusters_right;
-  std::unique_ptr<RowIDPosList> null_rows_left;
-  std::unique_ptr<RowIDPosList> null_rows_right;
+  MaterializedSegmentList<T> clusters_left;
+  MaterializedSegmentList<T> clusters_right;
+  RowIDPosList null_rows_left;
+  RowIDPosList null_rows_right;
 };
 
-/*
-*
-* Performs radix clustering for the sort merge join. The radix clustering algorithm clusters on the basis
-* of the least significant bits of the values because the values there are much more evenly distributed than for the
-* most significant bits. As a result, equal values always get moved to the same cluster and the clusters are
-* sorted in themselves but not in between the clusters. This is okay for the equi join, because we are only interested
-* in equality. In the case of a non-equi join however, complete sortedness is required, because join matches exist
-* beyond cluster borders. Therefore, the clustering defaults to a range clustering algorithm for the non-equi-join.
-* General clustering process:
-* -> Input chunks are materialized and sorted. Every value is stored together with its row id.
-* -> Then, either radix clustering or range clustering is performed.
-* -> At last, the resulting clusters are sorted.
-*
-* Radix clustering example:
-* cluster_count = 4
-* bits for 4 clusters: 2
-*
-*   000001|01
-*   000000|11
-*          ˆ right bits are used for clustering
-*
-**/
+// Performs radix clustering for the sort merge join. The radix clustering algorithm clusters on the basis of the least
+// significant bits of the values because the values there are much more evenly distributed than for the most
```

Comment:
We made this assumption back in 2018. I think the argument holds vs the most significant bits. I imagined a vector of values that all have a zero as the most significant bit because values are overall low. What happens if the data set does not match the assumption. Imagine a data set with only even numbers for example. Would it make sense to check for scenarios where radix clustering does not work well?

Reply:
Well, good question. Several things. First of all, we are still a very "benchmark-driven" research database and thus this assumption is fine for us (for now, we simply haven't seen "bad cases"). Production database regularly deal with data that is extremely skewed and radix partitioning is not going to solve that (e.g., single value accounts for 80% of your join column). It's interesting to look into that, but simply  not in our focus for now.
Second, if the even case would happen, we might lose half the partitions but that should not be too big of an impact. The heavily skewed case hurts much more.